blog/_posts/2021-03-25-kubernetes-on-ci.md

```diff
@@ -0,0 +1,150 @@
+up uses a one-time Kubernetes cluster to run tests against.
+
+A few of our friends have been asking us how we set it up and how to run their
+own clusters in CI. I've now explained it enough times that I should probably
+write down what we learned.
+
+Here are three ways to set it up, with the pros and cons of each!
+
+## Strategy #1: Local Cluster, Remote Registry
+
+Here's how I set up our first integration test framework.
+
+1. I created a dedicated gcr.io bucket for us to store images, and a GCP service
+  account with permission to write to it.
+
+2. I added the GCP service account credentials as a secret in our CI build.
+
+3. I forked [`kubeadm-dind-cluster`](https://github.com/kubernetes-retired/kubeadm-dind-cluster).
+  a set of Bash scripts to set up Kubernetes with Docker-in-Docker techniques.
+  
+All our test projects had Tilt build images, push them to the gcr.io bucket,
+then deploy servers that used these images.
+
+I barely got this working. A huge breakthrough! It caught so many subtle bugs
+and race conditions.
+
+I wouldn't call the Bash scripts _readable_. But they are hackable,
+cut-and-pasteable. There were examples of how to run it on CircleCI and
+TravisCI. `kubeadm-dind-cluster` has been deprecated in favor of more modern
+approaches like [`kind`](https://kind.sigs.k8s.io). But I learned a lot from its
+Bash scripts. We still use a lot of the techniques in this project today.
+
+There were other downsides though:
+
+- When drive-by contributors sent us PRs, the integration tests failed.
+  They didn't have access to to write to the gcr.io bucket. This made me so sad.
+  Contributors felt unwelcome. I never figured out a way to make this secure.
+  
+- We didn't reset the gcr.io bucket between test runs. So it was hard to guarantee
+  that images weren't leaking between tests. For example, if image pushing failed,
+  we wanted to be sure we weren't picking up a cached image from a previous test.
+
+## Strategy #2: Local Registry On a VM
+
+When I revisited this, I wanted to make sure:
+
+- Anyone could write to the image registry.
+
+- The image registry resets between runs.
+
+By this time, `kind` was taking off as the default choice for testing Kubernetes
+itself. `kind` also comes with the ability to run a local registry, so
+you can push images to the registry on `localhost:5000` and pull them from inside `kind`.
+
+I set up a new CI pipeline that:
+ 
+1. Creates a VM.
+
+2. Installs all the dependencies we needed, including Docker.
+
+3. Creates `kind` with a local registry, using their script.
+
+This worked well! And because the registry was local, it was faster too. We
+still use this approach to test `ctlptl` with both `minikube` and `kind`. Here's
+[the CI
+config](https://github.com/tilt-dev/ctlptl/blob/b6f808a09b05b6cf7aa0b3365e4781d2c23e4851/.circleci/config.yml#L30).
+
+But I weren't totally happy! VMs are slower. Upgrading dependencies is a bit
```

Comment:
```suggestion
But I wasn't totally happy! VMs are slower. Upgrading dependencies is a bit
```
(/we weren't)

Reply:
VMs are slower than what/upgrading deps is more heavyweight than what?

Reply:
done